### PR TITLE
add 3 retries to the file downloads

### PR DESCRIPTION
**Pull request by fearoffish**
To mitigate the flakey CDN problems we've been having, this adds a retry mechanism to `downloadFile`.

### CODE REVIEW

**Comment by robdimsdale** [comment]
+1 to using the exponential backoff library linked above. It works pretty well in all the situations we've tried it so far.

As far as the status codes for retrying, we had a conversation about whether there are certain status codes for which we should fail-fast, instead of retrying a request that will always fail. An example of this would be a `403` (invalid credentials). In principle, a user should not expect a `403` to succeed on retry, as the credentials are invalid. In practice, we often see some websites (like GitHub) returning `403`s when they should be returning other status codes (e.g. `429` too many requests, or `5xx` for an actual server issue).

Therefore, I am generally of the opinion that we shouldn't try to get too clever with failing-fast on some error codes - at least not until we have satisfied ourselves that the server is behaving appropriately. I'd rather err on the side of retrying unnecessarily than failing fast for a request that would have succeeded on retry.

**Comment by fearoffish** [comment]
Thanks for the feedback, I'll take a look at the exponentialBackup library now.

**Comment by fearoffish** [comment]
Question, @sophiewigmore, I can easily add the backoff using that library, which is nice...but testing it is another matter. I'll need to change the interface of the function (and all the places it's called from), to have a clock...otherwise the tests will be much slower. Is that okay, or should I not test the backoff?

**Comment by robdimsdale** [comment]
I think the way I would do this is add a `retryTimeLimit` (or `retryCount`, but I prefer time-limit because usually we care more about how long we retry for, not how many requests we make) to the `downloadFile` method, and propagate that all the way up to the [`Installer`](https://github.com/cloudfoundry/libbuildpack/blob/0a9b987be6a7b1d7af2aa60a1319abb87107aabb/installer.go#L16) struct. I would default it to something reasonable (like 10 seconds), and introduce a `SetRetryTimeout` method on the `Installer` to allow overriding. You wouldn't need to override the clock - you can just use real time values.

Then in the `installer_test.go` (which is the only place you'll actually test this, because the `downloadFile` method is unexported and isn't used elsewhere) you can configure a reasonable value for the override.

As far as testing, there are two paths:
1. In the happy-path case, where retry is not needed, it doesn't matter what the retry limits are because you won't hit them. So introducing retry won't slow down the test.
2. In the cases where you explicitly (or implicitly) test the retry behavior, you can set a fairly low value (e.g. 1 second) and while this will slow down the tests, it shouldn't be a significant impact to the developer.

Does that all make sense? [Here](https://github.com/paketo-buildpacks/github-config/blob/07e78c8c12cda62c283d9077649e702dbaca313c/actions/release/create/entrypoint/main_test.go#L764) is an example where we test a tool that uses the exponential backoff library that uses the patterns I describe above.

**Comment by fearoffish** [comment]
@robdimsdale I'm not 100% sure I understood, but I've gone with what I thought was reasonable.

I test that there is at least one retry, and I add two values (retry time limit and the initial interval of a retry) to the installer. Without both values I couldn't see how I'd test that retries actually work without the tests taking forever.

I'm open to thoughts though because I'm a new Golang developer and not in my comfort zone ;)

**Comment by sophiewigmore** [comment]
The tests look pretty good to me! 👏  I'll let Rob take a look also as he had a clearer vision for them

File: installer.go [hunk 1]
```diff
@@ -14,14 +14,16 @@ import (
...
+	appCacheDir              string
+	filesInAppCache          map[string]interface{}
+	versionLine              *map[string]string
+	retryTimeLimit           time.Duration
+	retryTimeInitialInterval time.Duration
 }
 
 func NewInstaller(manifest *Manifest) *Installer {
-	return &Installer{manifest, "", make(map[string]interface{}), &map[string]string{}}
+	return &Installer{manifest, "", make(map[string]interface{}), &map[string]string{}, 5 * time.Minute, 500 * time.Millisecond}
```
Comment (sophiewigmore) [inline installer.go:26]: I am wondering practically, if 5 minutes is too long for the retry window here. We've done 5 minutes in other code bases, but they exclusively run in our CI pipelines. We might want to shorten this down a bit possibly, so we have a slightly better user experience. Any thoughts?

File: installer_test.go [hunk 2]
```diff
@@ -213,6 +215,19 @@ var _ = Describe("Installer", func() {
 
 					Expect(outputFile).ToNot(BeAnExistingFile())
 				})
+				It("retries", func() {
```
[the remark this answers is not in the document]
Comment (fearoffish) [reply installer_test.go:251]: I liked the idea of a test for success after failure, so I added that too.

File: installer.go (hunk 1 above)
Comment (fearoffish) [reply installer.go]: I liked 1 minute as a timeout (if it's giving errors for a minute, that's enough) and an initial retry after 1 second. I've gone with that now. You're right, 5 minutes was too long.

File: installer_test.go [hunk 3]
```diff
@@ -190,6 +193,38 @@ var _ = Describe("Installer", func() {
...
+				It("makes intermediate directories", func() {
+					outputFile = filepath.Join(tmpdir, "notexist", "out.tgz")
+					err = installer.FetchDependency(inputs.Dependency, outputFile)
+
+					Expect(err).To(BeNil())
+					Expect(ioutil.ReadFile(outputFile)).To(Equal(inputs.ExpectedContent))
+				})
+
+				It("retries to get success", func() {
+					err = installer.FetchDependency(inputs.Dependency, outputFile)
```
Comment (robdimsdale) [inline installer_test.go:224]: Should we assert this error is `nil` like we do above?

File: util.go [hunk 4]
```diff
@@ -345,18 +347,43 @@ func CheckSha256(filePath, expectedSha256 string) error {
...
+	var resp *http.Response
+	var err error
+
+	// Define the operation to perform the HTTP request
+	operation := func() error {
+		resp, err = http.Get(url)
+
+		if err != nil {
+			return err
+			// return backoff.Permanent(err)
```
Comment (robdimsdale) [inline util.go:364]: I think we can remove this comment.

File: util.go [hunk 5]
```diff
@@ -345,18 +347,43 @@ func CheckSha256(filePath, expectedSha256 string) error {
...
+
+		if resp.StatusCode >= 400 {
+			return fmt.Errorf("%s", resp.Status)
+		}
+		return writeToFile(resp.Body, destFile, 0666)
+	}
+
+	// Define the notify function to handle retries
+	notify := func(err error, duration time.Duration) {
+		logger.Info("error: %v, retrying in %v...\n", err, duration)
```
Comment (robdimsdale) [inline util.go:376]: I think we can remove the `\n` from the log statement here.

**Review by sophiewigmore — APPROVED** [review]
Assuming Rob's last couple of comments get resolved, I'm happy with this as well :)

File: installer_test.go (hunk 3 above)
Comment (fearoffish) [reply installer_test.go:224]: we should

**Comment by fearoffish** [comment]
Fantastic, thanks to the both of you. My apologies for being a Golang beginner. I learnt plenty though and I think the final solution is better because of you both. Much appreciated.

**Review by robdimsdale — APPROVED** [review]
Awesome - thanks! I do not have the permissions to merge, but I assume @sophiewigmore does?

**Pull request merged**